cmd/juju/controller/register.go

```diff
@@ -233,74 +348,81 @@ func (c *registerCommand) maybeSetCurrentModel(ctx *cmd.Context, store jujuclien
 		if err != nil {
 			return errors.Trace(err)
 		}
-		fmt.Fprintf(ctx.Stderr, "\nCurrent model set to %q.\n\n", modelName)
-	} else {
-		fmt.Fprintf(ctx.Stderr, `
+		fmt.Fprintf(ctx.Stderr, "\nCurrent model set to %s.\n", modelName)
+		return nil
+	}
+	fmt.Fprintf(ctx.Stderr, `
 There are %d models available. Use "juju switch" to select
 one of them:
 `, len(models))
-		user := names.NewUserTag(userName)
-		ownerModelNames := make(set.Strings)
-		otherModelNames := make(set.Strings)
-		for _, model := range models {
-			if model.Owner == userName {
-				ownerModelNames.Add(model.Name)
-				continue
-			}
-			owner := names.NewUserTag(model.Owner)
-			modelName := common.OwnerQualifiedModelName(model.Name, owner, user)
-			otherModelNames.Add(modelName)
-		}
-		for _, modelName := range ownerModelNames.SortedValues() {
-			fmt.Fprintf(ctx.Stderr, "  - juju switch %s\n", modelName)
-		}
-		for _, modelName := range otherModelNames.SortedValues() {
-			fmt.Fprintf(ctx.Stderr, "  - juju switch %s\n", modelName)
+	user := names.NewUserTag(userName)
+	ownerModelNames := make(set.Strings)
+	otherModelNames := make(set.Strings)
+	for _, model := range models {
+		if model.Owner == userName {
+			ownerModelNames.Add(model.Name)
+			continue
 		}
-		fmt.Fprintln(ctx.Stderr)
+		owner := names.NewUserTag(model.Owner)
+		modelName := common.OwnerQualifiedModelName(model.Name, owner, user)
+		otherModelNames.Add(modelName)
+	}
+	for _, modelName := range ownerModelNames.SortedValues() {
+		fmt.Fprintf(ctx.Stderr, "  - juju switch %s\n", modelName)
+	}
+	for _, modelName := range otherModelNames.SortedValues() {
+		fmt.Fprintf(ctx.Stderr, "  - juju switch %s\n", modelName)
 	}
 	return nil
 }
 
 type registrationParams struct {
-	userTag         names.UserTag
-	controllerName  string
-	controllerAddrs []string
-	key             [32]byte
-	nonce           [24]byte
-	newPassword     string
+	// publicHost holds the host name of a public controller.
+	// If this is set, all other fields will be empty.
+	publicHost string
+
+	defaultControllerName string
+	userTag               names.UserTag
+	controllerAddrs       []string
+	key                   [32]byte
+	nonce                 [24]byte
+	newPassword           string
 }
 
 // getParameters gets all of the parameters required for registering, prompting
 // the user as necessary.
 func (c *registerCommand) getParameters(ctx *cmd.Context, store jujuclient.ClientStore) (*registrationParams, error) {
+	var params registrationParams
+	if strings.Contains(c.Arg, ".") || c.Arg == "localhost" {
+		// Looks like a host name - no URL-encoded base64 string should
+		// contain a dot and every public controller name should.
+		// Allow localhost for development purposes.
+		params.publicHost = c.Arg
```

Comment:
Yup, good catch - it's a hangover from when we fetched the controller name in both cases.